### PR TITLE
Add CustomResourceDefinition (WIP)

### PR DESCRIPTION
**Pull request by fhemberger**
Adds `CustomResourceDefinition`, which replaces `ThirdPartyResources` starting Kubernetes 1.7.

This is still work in progress, not sure I got your setup right. Tests are failing at the moment. Would be cool if you could give me a hand with this.

### CODE REVIEW

**Comment by silasbw** [comment]
@fhemberger thank you for starting on this -- we definitely need it.  I'll take a look tomorrow or the next day and see if I can help with the tests!

**Comment by fhemberger** [comment]
@silasbw Great, thank you. I got something wrong with the implementation but couldn't find out what it was … probably I'm just missing something.

Also CRDs can be namespace or cluster specific, I think a part of that is still missing.

File: test/api-extensions.test.js [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+        .get(`${ common.apiExtensions.path }/${ testCustomResourceName }`)
+        .reply(200, testApiExtensions);
+    });
+
+    // NOTE: Running only unit tests. Setting up CRD is more involved, and it
+    // makes it cumbersome to run the other integration tests. We need
+    // improvements to our integration test harness to make this work well.
+    common.only('unit', 'can POST and GET', done => {
+      async.series([
+        next => common.apiExtensions.post({ body: testApiExtensions }, next),
```
Comment (silasbw) [inline test/api-extensions.test.js]: You want this:

```js
next => common.apiExtensions.customresourcedefinitions.post({ body: testApiExtensions }, next),
next => common.apiExtensions.customresourcedefinitions.get(testCustomResourceName, next)
```

https://kubernetes.io/docs/api-reference/v1.8/#-strong-write-operations-strong--254

File: lib/api-extensions.js [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+const ApiGroup = require('./api-group');
+const CustomResourceDefiniton = require('./custom-resource-definition');
+
+class ApiExtensions extends ApiGroup {
+  constructor(options) {
+    options = Object.assign({}, options, {
+      path: 'apis/apiextensions.k8s.io',
+      version: options.version || 'v1beta1',
+      groupResources: [
+        'customresourcedefinition'
```
Comment (silasbw) [inline lib/api-extensions.js]: pluralize: `customresourcedefinitions`

The code automatically creates a singular alias so `apiExtensions.customresourcedefinitions` and `apiExtensions.customresourcedefinition` will both work, but you need to specify the plural version here.

File: lib/api-extensions.js [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+      ],
+      namespaceResources: [
+        { name: 'customresourcedefinition', Constructor: CustomResourceDefiniton }
```
Comment (silasbw) [inline lib/api-extensions.js]: pluralize here too.

File: test/api-extensions.test.js [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+  }
+};
+
+describe('lib.apiextensions', () => {
+  describe('.ApiExtensions', () => {
+    const testCustomResourceName = testApiExtensions.metadata.name;
+
+    beforeTesting('unit', () => {
+      nock(common.apiExtensions.url)
+        .post(`${ common.apiExtensions.path }`)
```
Comment (silasbw) [inline test/api-extensions.test.js]: you'll need to update these mocks:

```js
        .post(`${ common.apiExtensions.path }/customresourcedefinitions`)
        .reply(201, testApiExtensions)
        .get(`${ common.apiExtensions.path }/customresourcedefinitions/${ testCustomResourceName }`)
        .reply(200, testApiExtensions);
```

File: test/common.js [hunk 5]
```diff
@@ -76,6 +79,9 @@ function injectApis(options) {
     rbac: { Constructor: Rbac },
     thirdPartyResources: {
       Constructor: ThirdPartyResources, options: { group: 'kubernetes-client.com' }
+    },
+    customResourceDefinition: {
```
Comment (silasbw) [inline test/common.js]: pluralize (similar to `thirdPartyResources`)

**Comment by fhemberger** [comment]
Thanks, but now 4 instead of two tests are failing … :thinking: 
I need to dig deeper into it.

**Comment by silasbw** [comment]
Hey @fhemberger sent you https://github.com/fhemberger/kubernetes-client/pull/1.

unit tests work, integration tests have some issues.  I'll take a look at those when i get a moment.

**Comment by fhemberger** [comment]
Ok, back from vacation, updated PR with your changes, @silasbw.
Are we good to go or ist there still something missing?

**Comment by silasbw** [comment]
Hey @fhemberger, I sent you some tweaks to fix integration tests, etc:

  https://github.com/fhemberger/kubernetes-client/pull/2

Sorry this has been such a slog and thanks for sticking with it.  We're working on making these types of contributions much easier to make.

**Comment by fhemberger** [comment]
@silasbw Awesome, tests are passing! :100:

**Comment by silasbw** [comment]
I published these changes in `kubernetes-client@3.17.0`.  Thanks again.

**Pull request merged**